### ISSUE TITLE
A lint should prevent the use of `assert!()`

### ISSUE THREAD

**Issue by NthPortal** [rust-lang/rust-clippy#16346]
### Summary

The changes suggested by `manual_assert` (that is, switching to `assert!`) cause the code that panics to no longer trigger the `panic` lint.

### Lint Name

`panic`

### Reproducer

I tried this code:

```rust
#[expect(clippy::panic, reason = "allowed")]
#[expect(clippy::assertions_on_constants, reason = "simple repro")]
fn main() {
    assert!(false, "oh no")
}
```

([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=c2395c8d90e64b5323e199f6e750d08e))

I expected it to compile with no warnings.

Instead, this happened:

```
warning: this lint expectation is unfulfilled
 --> src/main.rs:1:10
  |
1 | #[expect(clippy::panic, reason = "allowed")]
  |          ^^^^^^^^^^^^^
  |
  = note: allowed
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```

In comparison, the following matches the `panic` lint but also the `manual_assert` lint:

```rust
#[expect(clippy::panic, reason = "allowed")]
fn main() {
    if true {
        panic!("oh no");
    }
}
```

([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=96d3b2c1410cbe0be401d6e7ca141119))

### Version

```text
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

**Comment by samueltardieu** [comment]
Just to be clear: your complaint is that `clippy::panic` will not trigger in the following code, right?

```rust
#[expect(clippy::panic, reason = "allowed")]
#[expect(clippy::assertions_on_constants, reason = "simple repro")]
fn main() {
    assert!(false, "oh no")
}
```

If so, we can relabel the issue to not add `clippy::manual_assert` to the mix as it might be confusing.

**Comment by NthPortal** [comment]
> your complaint is that `clippy::panic` will not trigger in the following code

yeah, having `assert!` trigger `clippy::panic` is the outcome I'd prefer, but I wasn't sure if there was a specific reason it didn't

**Comment by samueltardieu** [comment]
Well, that would have to be another lint then, as `clippy::panic` triggers only on explicit `panic!()` use.

@rustbot label -C-bug -I-false-negative C-feature-request

**Comment by ada4a** [comment]
You might be interested in [`disallowed_macros`](https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_macros)?

**Comment by NthPortal** [comment]
@ada4a while that could potentially be a workaround, the issue is that I want to forbid things that panic, not that I object to it as a macro. `panic`, `unwrap_used` and `expect_used` are already lints; it feels like `assert!` should have one too.

@samueltardieu a thought I just had: an `assert` lint probably ought to match `assert_eq!` and `assert_ne!` as well
